types/src/main/scala/roc/types/decoders.scala

```diff
@@ -196,24 +289,73 @@ object decoders {
 
   implicit val zonedDateTimeElementDecoders: ElementDecoder[TimestampWithTZ] = 
     new ElementDecoder[TimestampWithTZ] {
+
+      private val POSTGRES_TIMESTAMP_EPOCH =
+        OffsetDateTime.of(2000, 1, 1, 0, 0, 0, 0, ZoneOffset.UTC).toInstant.toEpochMilli
+
+      //Postgres uses milliseconds since its own custom epoch (see above)
+      private def timestampToInstant(microseconds: Long) =
+        Instant.ofEpochMilli(microseconds / 1000L).plusMillis(POSTGRES_TIMESTAMP_EPOCH)
+
       private val zonedDateTimeFmt = new DateTimeFormatterBuilder()
         .appendPattern("yyyy-MM-dd HH:mm:ss")
         .appendFraction(ChronoField.MICRO_OF_SECOND, 0, 6, true)
         .appendOptional(DateTimeFormatter.ofPattern("X"))
         .toFormatter()
+
       def textDecoder(text: String): TimestampWithTZ = Xor.catchNonFatal({
         ZonedDateTime.parse(text, zonedDateTimeFmt)
       }).fold(
         {l => throw new ElementDecodingFailure("TIMESTAMP WITH TIME ZONE", l)},
         {r => r}
       )
+
       def binaryDecoder(bytes: Array[Byte]): TimestampWithTZ = Xor.catchNonFatal({
-        val text = new String(bytes, StandardCharsets.UTF_8)
-        ZonedDateTime.parse(text, zonedDateTimeFmt)
+        val buf = ByteBuffer.wrap(bytes)
+        timestampToInstant(buf.getLong).atZone(ZoneId.systemDefault())
       }).fold(
         {l => throw new ElementDecodingFailure("TIMESTAMP WITH TIME ZONE", l)},
         {r => r}
       )
       def nullDecoder: TimestampWithTZ = throw new NullDecodedFailure("TIMSTAMP WITH TIME ZONE")
     }
+
+  implicit val hstoreDecoders: ElementDecoder[HStore] = new ElementDecoder[HStore] {
+    val hstoreStringRegex = """"([^"]*)"=>(?:NULL|"([^"]*))"""".r
+    def textDecoder(text: String): HStore = Xor.catchNonFatal {
+      hstoreStringRegex.findAllMatchIn(text).map {
+        m => m.group(1) -> Option(m.group(2))
+      }.toMap
+    }.fold(
+      l => throw new ElementDecodingFailure("HSTORE", l),
+      identity
+    )
+
+    def nullDecoder(): HStore = throw new NullDecodedFailure("HSTORE")
+
+    def binaryDecoder(bytes: Array[Byte]): HStore = {
+      val buf = ByteBuffer.wrap(bytes)
+      Xor.catchNonFatal {
+        val count = buf.getInt()
+        val charset = StandardCharsets.UTF_8 //TODO: are we going to support other charsets?
```

Comment:
Yeah, most of the current machines were written after not so much a copy paste, but a heavy look over from that code base.